File: frontend/src/api/market.ts

import { apiGet } from "./client";

export type Bar = {
  bar_at: string;
  open: number | null;
  high: number | null;
  low: number | null;
  close: number;
  volume: number | null;
};

export type BarsResponse = {
  symbol: string;
  timeframe: string;
  bars: Bar[];
};

export type Quote = {
  symbol: string;
  price: number;
  as_of: string | null;
  updated_at: string;
};

export type OrderMarker = {
  id: string;
  side: string;
  status: string;
  quantity: number | null;
  notional: number | null;
  filled_at: string | null;
  submitted_at: string | null;
  filled_price: number | null;
  stop_loss: Record<string, unknown> | null;
  take_profit: Record<string, unknown> | null;
};

export type OrderMarkersResponse = {
  symbol: string;
  orders: OrderMarker[];
};

export type ProposalMarker = {
  id: string;
  strategy_id: string | null;
  outcome: string;
  confidence: number | null;
  market_data_timestamp: string | null;
  reasoning_text: string | null;
  created_at: string;
};

export type RiskEventMarker = {
  id: string;
  agent_decision_id: string;
  outcome: string;
  reasons: unknown[];
  market_data_timestamp: string | null;
  created_at: string;
};

export type DecisionMarkersResponse = {
  symbol: string;
  proposals: ProposalMarker[];
  risk_events: RiskEventMarker[];
};

export type StrategyActivity = {
  strategy_id: string;
  type_code: string;
  name: string;
  status: string;
  order_count: number;
  buy_count: number;
  sell_count: number;
  total_notional: number;
};

export type StrategyActivityResponse = {
  strategies: StrategyActivity[];
};

export async function fetchSymbols(): Promise<string[]> {
  const res = await apiGet<{ symbols: string[] }>("/api/market/symbols");
  return res.symbols;
}

export async function fetchBars(symbol: string, timeframe = "1Day", limit = 200): Promise<BarsResponse> {
  const params = new URLSearchParams({ symbol, timeframe, limit: String(limit) });
  return apiGet<BarsResponse>(`/api/market/bars?${params.toString()}`);
}

export async function fetchQuote(symbol: string): Promise<Quote> {
  const params = new URLSearchParams({ symbol });
  return apiGet<Quote>(`/api/market/quote?${params.toString()}`);
}

export async function fetchOrderMarkers(symbol: string, limit = 100): Promise<OrderMarkersResponse> {
  const params = new URLSearchParams({ symbol, limit: String(limit) });
  return apiGet<OrderMarkersResponse>(`/api/market/orders?${params.toString()}`);
}

export async function fetchDecisionMarkers(symbol: string, limit = 50): Promise<DecisionMarkersResponse> {
  const params = new URLSearchParams({ symbol, limit: String(limit) });
  return apiGet<DecisionMarkersResponse>(`/api/market/decisions?${params.toString()}`);
}

export async function fetchStrategyActivity(): Promise<StrategyActivityResponse> {
  return apiGet<StrategyActivityResponse>("/api/market/strategy-activity");
}

File: backend/app/schemas/market.py
"""Schémas de lecture des données de marché et des marqueurs de graphique
(§B27 "Graphiques marché et analytics") — voir `backend/app/market.py` pour
la justification complète de chaque source de données."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel


class SymbolsResponse(BaseModel):
    symbols: list[str]


class BarOut(BaseModel):
    bar_at: datetime
    open: float | None
    high: float | None
    low: float | None
    close: float
    volume: float | None

    model_config = {"from_attributes": True}


class BarsResponse(BaseModel):
    symbol: str
    timeframe: str
    bars: list[BarOut]


class QuoteOut(BaseModel):
    symbol: str
    price: float
    as_of: datetime | None
    updated_at: datetime

    model_config = {"from_attributes": True}


class OrderMarkerOut(BaseModel):
    id: uuid.UUID
    side: str
    status: str
    quantity: float | None
    notional: float | None
    filled_at: datetime | None
    submitted_at: datetime | None
    filled_price: float | None
    stop_loss: dict | None
    take_profit: dict | None


class OrderMarkersResponse(BaseModel):
    symbol: str
    orders: list[OrderMarkerOut]


class ProposalMarkerOut(BaseModel):
    id: uuid.UUID
    strategy_id: uuid.UUID | None
    outcome: str
    confidence: int | None
    market_data_timestamp: str | None
    reasoning_text: str | None
    created_at: datetime


class RiskEventMarkerOut(BaseModel):
    id: uuid.UUID
    agent_decision_id: uuid.UUID
    outcome: str
    reasons: list
    market_data_timestamp: str | None
    created_at: datetime


class DecisionMarkersResponse(BaseModel):
    symbol: str
    proposals: list[ProposalMarkerOut]
    risk_events: list[RiskEventMarkerOut]


class StrategyActivityOut(BaseModel):
    strategy_id: uuid.UUID
    type_code: str
    name: str
    status: str
    order_count: int
    buy_count: int
    sell_count: int
    total_notional: float


class StrategyActivityResponse(BaseModel):
    # §B27 "Performance par stratégie" — voir docstring de
    # `market.strategy_activity` : proxy honnête (activité d'ordres réels),
    # pas un P&L attribué par stratégie (non suivi à ce jour).
    strategies: list[StrategyActivityOut]
